{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6d377ce1",
   "metadata": {},
   "source": [
    "# EzyPredict - prepare protein data\n",
    "\n",
    "### Predicting enzyme function from protein sequence, structure and physicochemical properties\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9c811d3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install dependencies\n",
    "#!pip install cobra\n",
    "#!pip install requests\n",
    "#!pip install fair-esm\n",
    "#!pip install torch\n",
    "#!pip install tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cfc12e11",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import requests\n",
    "import torch\n",
    "from EzyPredict_helperfunctions import *\n",
    "from cobra import io\n",
    "import tqdm\n",
    "import esm\n",
    "import pickle\n",
    "import subprocess\n",
    "import re\n",
    "import esm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1db4c895",
   "metadata": {},
   "source": [
    "## Define a set of enzymes\n",
    "\n",
    "I don't want to predict the function of all types of proteins - just of enzymes. So, I need to select just enzymes from whole genomes. To start with, because I only have a local computer to train my model on, I will just extract all enzymes that are part of the Sachharomyces cerevisiae metabolic model.\n",
    "\n",
    "I will do this using the latest genome-scale metabolic model of yeast - Yeast8.0 (https://github.com/SysBioChalmers/yeast-GEM) using the cobratoolbox (https://opencobra.github.io/cobratoolbox/stable/tutorials/) implemented in python as cobrapy (https://github.com/opencobra/cobrapy).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "417802dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "model_dir = \"yeast-GEM/model\"\n",
    "model_file = next(f for f in os.listdir(model_dir) if f.endswith('.xml'))  # find the first .xml file in the directory\n",
    "model_path = os.path.join(model_dir, model_file)\n",
    "\n",
    "# Load the model\n",
    "model = io.read_sbml_model(model_path)\n",
    "\n",
    "enzymes = []\n",
    "\n",
    "for reaction in model.reactions:\n",
    "    # Get the list of enzymes (gene products) catalyzing the reaction\n",
    "    enzymes.extend([gene.id for gene in reaction.genes])\n",
    "\n",
    "# Remove duplicates by converting the list to a set and then back to a list\n",
    "enzymes = list(set(enzymes))\n",
    "#print(enzyme[1:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a223e33",
   "metadata": {},
   "source": [
    "## Get sequences from UniProt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c515d077",
   "metadata": {},
   "source": [
    "Uniprot doesn't take ORF Ids that are stored in 'enzymes' we need to convert these to UniProt identifiers. I will do this using a mapping file that was downloaded from UniProt.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ce7fa2a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Convert ORF to UniProt\n",
    "uniprot_ids = [convert_ORF2Uniprot(enzyme) for enzyme in enzymes]\n",
    "#print(uniprot_ids[1:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "613f5450",
   "metadata": {},
   "source": [
    "## ESM2 encodings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54510ebd",
   "metadata": {},
   "source": [
    "### Load the ESM2 model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "74c2c93a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## this downloads the ESM1b language model which is a 7GB file. Don't run more than once\n",
    "# provided as a .zip in the repo\n",
    "model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S() \n",
    "batch_converter = alphabet.get_batch_converter()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828543fb",
   "metadata": {},
   "source": [
    "## Fetch the sequence of each uniprotID and encode it\n",
    "\n",
    "Using helper functions in EzyPredict_helperfunctions.py, I will pass every UniProtID of the enzymes I collected first to UniProt API and get an amino acid sequence. Then, I will pass it to ESM2 to get a sequence encoding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "661f8976",
   "metadata": {},
   "outputs": [],
   "source": [
    "## ESM2 doesn't accept sequences above 1024aa.\n",
    "# I don't know what to do about this yet so i'm just trimming the sequences for now \n",
    "\n",
    "def encode_sequence_with_ESM2_trim(sequence):\n",
    "    # If sequence is too long, trim it from both sides\n",
    "    max_sequence_length = 1024 \n",
    "    was_trimmed = False\n",
    "    while len(sequence) > max_sequence_length:\n",
    "        was_trimmed = True\n",
    "        excess_length = len(sequence) - max_sequence_length\n",
    "        trim_amount = (excess_length - 1) // 2  \n",
    "        sequence = sequence[trim_amount:-trim_amount]\n",
    "        \n",
    "    # Prepare the data for ESM2\n",
    "    data = [(sequence, sequence)] \n",
    "    batch_labels, batch_strs, batch_tokens = batch_converter(data)\n",
    "    \n",
    "    # Get the representations.\n",
    "    with torch.no_grad():\n",
    "        results = model(batch_tokens, repr_layers=[33])\n",
    "        token_representations = results[\"representations\"][33]\n",
    "    \n",
    "    avg_representation = token_representations.mean(1)\n",
    "    \n",
    "    return avg_representation, was_trimmed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9e66bd2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Example of what the encodings look like \n",
    "#encoded_representations = {}\n",
    "\n",
    "#for uid in uniprot_ids[0:3]:\n",
    " #   sequence = fetch_sequence_from_uniprot(uid)\n",
    "  #  if sequence:\n",
    "   #     encoding = encode_sequence_with_ESM2_trim(sequence)\n",
    "    #    encoded_representations[uid] = encoding\n",
    "#print(encoded_representations)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab42732d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1163/1163 [21:07<00:00,  1.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "70 sequences were trimmed due to exceeding the model's limit.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "def save_to_disk(data, filename=\"ESM2_encoded_representations.pkl\"):\n",
    "    with open(filename, \"wb\") as f: \n",
    "        pickle.dump(data, f)\n",
    "\n",
    "encoded_representations = {}\n",
    "trimmed_count = 0 \n",
    "\n",
    "# Loop with progress bar\n",
    "for uid in tqdm.tqdm(uniprot_ids):\n",
    "    sequence = fetch_sequence_from_uniprot(uid)\n",
    "    if sequence:\n",
    "        encoding, was_trimmed = encode_sequence_with_ESM2_trim(sequence)\n",
    "        if not was_trimmed: ## I'm only saving the untrimmed because PCA later on showed that this was messing up the data\n",
    "            encoded_representations[uid] = encoding\n",
    "        else:\n",
    "            trimmed_count += 1\n",
    "\n",
    "# \n",
    "save_to_disk(encoded_representations)\n",
    "\n",
    "# Print the number of trimmed sequences\n",
    "print(f\"{trimmed_count} sequences were trimmed due to exceeding the model's limit.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cf05c8d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "70 sequences were trimmed.\n"
     ]
    }
   ],
   "source": [
    "# Count how many sequences were trimmed\n",
    "print(f\"{trimmed_count} sequences were trimmed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8cddd4d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1093\n"
     ]
    }
   ],
   "source": [
    "# count how many sequences we have encoded\n",
    "with open(\"ESM2_encoded_representations.pkl\", \"rb\") as f:\n",
    "    encoded_representations = pickle.load(f)\n",
    "print(len(encoded_representations))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d376ee2d",
   "metadata": {},
   "source": [
    "## Pepstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa42905b",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_directory = os.getcwd()\n",
    "directory = \"pepstats_results\"\n",
    "import os\n",
    "if not os.path.exists(directory):\n",
    "    os.makedirs(directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70c49175",
   "metadata": {},
   "source": [
    "### loop over all uniprot IDs, fetch sequence from UniProt API, then get pepstats from PEPSTATS API"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29018896",
   "metadata": {},
   "source": [
    "## no need to run this again - all results are in pepstats_results\n",
    "## Turn this into a Code cell if you want to run the fetching from pepstats api otherwise load the results \n",
    "\n",
    "for uid in tqdm.tqdm(uniprot_ids):\n",
    "    \n",
    "    # fetch sequence from UniProt\n",
    "    sequence = fetch_sequence_from_uniprot(uid)\n",
    "    \n",
    "    # ensure sequence was successfully fetched\n",
    "    if not sequence:\n",
    "        print(f\"Failed to fetch sequence for {uid}\")\n",
    "        continue\n",
    "    \n",
    "    # Submit the sequence to pepstats\n",
    "    job_id = submit_sequence(sequence)\n",
    "    \n",
    "    # Keep checking every 30 seconds if the job is done\n",
    "    while True:\n",
    "        if is_job_done(job_id):\n",
    "            break\n",
    "        time.sleep(30)\n",
    "    \n",
    "    # Download results once the job is done\n",
    "    filename = f\"{uid}_pepstats\"\n",
    "    output_path = os.path.join(directory, filename)\n",
    "    retrieve_job_results(job_id, output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1f441b5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isoelectric Point</th>\n",
       "      <th>Tiny</th>\n",
       "      <th>Small</th>\n",
       "      <th>Aliphatic</th>\n",
       "      <th>Aromatic</th>\n",
       "      <th>Non-polar</th>\n",
       "      <th>Polar</th>\n",
       "      <th>Charged</th>\n",
       "      <th>Basic</th>\n",
       "      <th>Acidic</th>\n",
       "      <th>Total Residues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P48813</th>\n",
       "      <td>7.7144</td>\n",
       "      <td>202</td>\n",
       "      <td>331</td>\n",
       "      <td>211</td>\n",
       "      <td>87</td>\n",
       "      <td>397</td>\n",
       "      <td>266</td>\n",
       "      <td>134</td>\n",
       "      <td>73</td>\n",
       "      <td>61</td>\n",
       "      <td>663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53241</th>\n",
       "      <td>9.0584</td>\n",
       "      <td>161</td>\n",
       "      <td>274</td>\n",
       "      <td>171</td>\n",
       "      <td>113</td>\n",
       "      <td>348</td>\n",
       "      <td>245</td>\n",
       "      <td>127</td>\n",
       "      <td>76</td>\n",
       "      <td>51</td>\n",
       "      <td>593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P33550</th>\n",
       "      <td>5.5352</td>\n",
       "      <td>95</td>\n",
       "      <td>179</td>\n",
       "      <td>94</td>\n",
       "      <td>79</td>\n",
       "      <td>220</td>\n",
       "      <td>205</td>\n",
       "      <td>115</td>\n",
       "      <td>57</td>\n",
       "      <td>58</td>\n",
       "      <td>425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q02207</th>\n",
       "      <td>9.4507</td>\n",
       "      <td>271</td>\n",
       "      <td>469</td>\n",
       "      <td>283</td>\n",
       "      <td>102</td>\n",
       "      <td>496</td>\n",
       "      <td>404</td>\n",
       "      <td>213</td>\n",
       "      <td>123</td>\n",
       "      <td>90</td>\n",
       "      <td>900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P21375</th>\n",
       "      <td>9.4952</td>\n",
       "      <td>145</td>\n",
       "      <td>261</td>\n",
       "      <td>149</td>\n",
       "      <td>42</td>\n",
       "      <td>250</td>\n",
       "      <td>251</td>\n",
       "      <td>124</td>\n",
       "      <td>72</td>\n",
       "      <td>52</td>\n",
       "      <td>501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P38827</th>\n",
       "      <td>9.3539</td>\n",
       "      <td>275</td>\n",
       "      <td>492</td>\n",
       "      <td>253</td>\n",
       "      <td>123</td>\n",
       "      <td>466</td>\n",
       "      <td>614</td>\n",
       "      <td>351</td>\n",
       "      <td>204</td>\n",
       "      <td>147</td>\n",
       "      <td>1080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53871</th>\n",
       "      <td>7.2464</td>\n",
       "      <td>103</td>\n",
       "      <td>179</td>\n",
       "      <td>99</td>\n",
       "      <td>43</td>\n",
       "      <td>182</td>\n",
       "      <td>175</td>\n",
       "      <td>98</td>\n",
       "      <td>54</td>\n",
       "      <td>44</td>\n",
       "      <td>357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0CH64</th>\n",
       "      <td>5.9257</td>\n",
       "      <td>66</td>\n",
       "      <td>123</td>\n",
       "      <td>71</td>\n",
       "      <td>27</td>\n",
       "      <td>121</td>\n",
       "      <td>104</td>\n",
       "      <td>51</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P32622</th>\n",
       "      <td>5.5463</td>\n",
       "      <td>135</td>\n",
       "      <td>254</td>\n",
       "      <td>138</td>\n",
       "      <td>50</td>\n",
       "      <td>243</td>\n",
       "      <td>252</td>\n",
       "      <td>135</td>\n",
       "      <td>67</td>\n",
       "      <td>68</td>\n",
       "      <td>495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q03175</th>\n",
       "      <td>10.0252</td>\n",
       "      <td>74</td>\n",
       "      <td>141</td>\n",
       "      <td>91</td>\n",
       "      <td>51</td>\n",
       "      <td>178</td>\n",
       "      <td>165</td>\n",
       "      <td>95</td>\n",
       "      <td>60</td>\n",
       "      <td>35</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1163 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Isoelectric Point Tiny Small Aliphatic Aromatic Non-polar Polar  \\\n",
       "P48813            7.7144  202   331       211       87       397   266   \n",
       "P53241            9.0584  161   274       171      113       348   245   \n",
       "P33550            5.5352   95   179        94       79       220   205   \n",
       "Q02207            9.4507  271   469       283      102       496   404   \n",
       "P21375            9.4952  145   261       149       42       250   251   \n",
       "...                  ...  ...   ...       ...      ...       ...   ...   \n",
       "P38827            9.3539  275   492       253      123       466   614   \n",
       "P53871            7.2464  103   179        99       43       182   175   \n",
       "P0CH64            5.9257   66   123        71       27       121   104   \n",
       "P32622            5.5463  135   254       138       50       243   252   \n",
       "Q03175           10.0252   74   141        91       51       178   165   \n",
       "\n",
       "       Charged Basic Acidic  Total Residues  \n",
       "P48813     134    73     61             663  \n",
       "P53241     127    76     51             593  \n",
       "P33550     115    57     58             425  \n",
       "Q02207     213   123     90             900  \n",
       "P21375     124    72     52             501  \n",
       "...        ...   ...    ...             ...  \n",
       "P38827     351   204    147            1080  \n",
       "P53871      98    54     44             357  \n",
       "P0CH64      51    26     25             225  \n",
       "P32622     135    67     68             495  \n",
       "Q03175      95    60     35             343  \n",
       "\n",
       "[1163 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# After processing all sequences, collect results into a dataframe\n",
    "pepstats_df = collate_pepstats_to_dataframe(directory)\n",
    "\n",
    "# drop the columns that are not relevant for function and create big separation in PCA\n",
    "pepstats_df = pepstats_df.drop(columns=[\"Molecular weight\", \"Average Residue Weight\",\n",
    "                                       \"Improbability of expression in inclusion bodies\",\n",
    "                                       \"A280 Molar Extinction Coefficients\"])\n",
    "\n",
    "#pepstats_df\n",
    "pepstats_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dcffd3b",
   "metadata": {},
   "source": [
    "#### Calculate fraction of Tiny, Small, Aliphate etc amino acids.. the total number is not important so after calculating the fraction , we remove it "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "50e5b37a",
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "columns_to_divide = [\"Tiny\", \"Small\", \"Aliphatic\", \"Aromatic\", \"Non-polar\", \"Polar\", \"Charged\", \"Basic\", \"Acidic\"]\n",
    "\n",
    "for column in columns_to_divide:\n",
    "    pepstats_df[column] = pepstats_df[column].astype(float) / pepstats_df[\"Total Residues\"]\n",
    "pepstats_df = pepstats_df.drop('Total Residues', axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5737e563",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isoelectric Point</th>\n",
       "      <th>Tiny</th>\n",
       "      <th>Small</th>\n",
       "      <th>Aliphatic</th>\n",
       "      <th>Aromatic</th>\n",
       "      <th>Non-polar</th>\n",
       "      <th>Polar</th>\n",
       "      <th>Charged</th>\n",
       "      <th>Basic</th>\n",
       "      <th>Acidic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P48813</th>\n",
       "      <td>7.7144</td>\n",
       "      <td>0.304676</td>\n",
       "      <td>0.499246</td>\n",
       "      <td>0.318250</td>\n",
       "      <td>0.131222</td>\n",
       "      <td>0.598793</td>\n",
       "      <td>0.401207</td>\n",
       "      <td>0.202112</td>\n",
       "      <td>0.110106</td>\n",
       "      <td>0.092006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53241</th>\n",
       "      <td>9.0584</td>\n",
       "      <td>0.271501</td>\n",
       "      <td>0.462057</td>\n",
       "      <td>0.288364</td>\n",
       "      <td>0.190556</td>\n",
       "      <td>0.586847</td>\n",
       "      <td>0.413153</td>\n",
       "      <td>0.214165</td>\n",
       "      <td>0.128162</td>\n",
       "      <td>0.086003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P33550</th>\n",
       "      <td>5.5352</td>\n",
       "      <td>0.223529</td>\n",
       "      <td>0.421176</td>\n",
       "      <td>0.221176</td>\n",
       "      <td>0.185882</td>\n",
       "      <td>0.517647</td>\n",
       "      <td>0.482353</td>\n",
       "      <td>0.270588</td>\n",
       "      <td>0.134118</td>\n",
       "      <td>0.136471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q02207</th>\n",
       "      <td>9.4507</td>\n",
       "      <td>0.301111</td>\n",
       "      <td>0.521111</td>\n",
       "      <td>0.314444</td>\n",
       "      <td>0.113333</td>\n",
       "      <td>0.551111</td>\n",
       "      <td>0.448889</td>\n",
       "      <td>0.236667</td>\n",
       "      <td>0.136667</td>\n",
       "      <td>0.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P21375</th>\n",
       "      <td>9.4952</td>\n",
       "      <td>0.289421</td>\n",
       "      <td>0.520958</td>\n",
       "      <td>0.297405</td>\n",
       "      <td>0.083832</td>\n",
       "      <td>0.499002</td>\n",
       "      <td>0.500998</td>\n",
       "      <td>0.247505</td>\n",
       "      <td>0.143713</td>\n",
       "      <td>0.103792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P38827</th>\n",
       "      <td>9.3539</td>\n",
       "      <td>0.254630</td>\n",
       "      <td>0.455556</td>\n",
       "      <td>0.234259</td>\n",
       "      <td>0.113889</td>\n",
       "      <td>0.431481</td>\n",
       "      <td>0.568519</td>\n",
       "      <td>0.325000</td>\n",
       "      <td>0.188889</td>\n",
       "      <td>0.136111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53871</th>\n",
       "      <td>7.2464</td>\n",
       "      <td>0.288515</td>\n",
       "      <td>0.501401</td>\n",
       "      <td>0.277311</td>\n",
       "      <td>0.120448</td>\n",
       "      <td>0.509804</td>\n",
       "      <td>0.490196</td>\n",
       "      <td>0.274510</td>\n",
       "      <td>0.151261</td>\n",
       "      <td>0.123249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0CH64</th>\n",
       "      <td>5.9257</td>\n",
       "      <td>0.293333</td>\n",
       "      <td>0.546667</td>\n",
       "      <td>0.315556</td>\n",
       "      <td>0.120000</td>\n",
       "      <td>0.537778</td>\n",
       "      <td>0.462222</td>\n",
       "      <td>0.226667</td>\n",
       "      <td>0.115556</td>\n",
       "      <td>0.111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P32622</th>\n",
       "      <td>5.5463</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.513131</td>\n",
       "      <td>0.278788</td>\n",
       "      <td>0.101010</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>0.509091</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.135354</td>\n",
       "      <td>0.137374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q03175</th>\n",
       "      <td>10.0252</td>\n",
       "      <td>0.215743</td>\n",
       "      <td>0.411079</td>\n",
       "      <td>0.265306</td>\n",
       "      <td>0.148688</td>\n",
       "      <td>0.518950</td>\n",
       "      <td>0.481050</td>\n",
       "      <td>0.276968</td>\n",
       "      <td>0.174927</td>\n",
       "      <td>0.102041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1163 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Isoelectric Point      Tiny     Small  Aliphatic  Aromatic  Non-polar  \\\n",
       "P48813            7.7144  0.304676  0.499246   0.318250  0.131222   0.598793   \n",
       "P53241            9.0584  0.271501  0.462057   0.288364  0.190556   0.586847   \n",
       "P33550            5.5352  0.223529  0.421176   0.221176  0.185882   0.517647   \n",
       "Q02207            9.4507  0.301111  0.521111   0.314444  0.113333   0.551111   \n",
       "P21375            9.4952  0.289421  0.520958   0.297405  0.083832   0.499002   \n",
       "...                  ...       ...       ...        ...       ...        ...   \n",
       "P38827            9.3539  0.254630  0.455556   0.234259  0.113889   0.431481   \n",
       "P53871            7.2464  0.288515  0.501401   0.277311  0.120448   0.509804   \n",
       "P0CH64            5.9257  0.293333  0.546667   0.315556  0.120000   0.537778   \n",
       "P32622            5.5463  0.272727  0.513131   0.278788  0.101010   0.490909   \n",
       "Q03175           10.0252  0.215743  0.411079   0.265306  0.148688   0.518950   \n",
       "\n",
       "           Polar   Charged     Basic    Acidic  \n",
       "P48813  0.401207  0.202112  0.110106  0.092006  \n",
       "P53241  0.413153  0.214165  0.128162  0.086003  \n",
       "P33550  0.482353  0.270588  0.134118  0.136471  \n",
       "Q02207  0.448889  0.236667  0.136667  0.100000  \n",
       "P21375  0.500998  0.247505  0.143713  0.103792  \n",
       "...          ...       ...       ...       ...  \n",
       "P38827  0.568519  0.325000  0.188889  0.136111  \n",
       "P53871  0.490196  0.274510  0.151261  0.123249  \n",
       "P0CH64  0.462222  0.226667  0.115556  0.111111  \n",
       "P32622  0.509091  0.272727  0.135354  0.137374  \n",
       "Q03175  0.481050  0.276968  0.174927  0.102041  \n",
       "\n",
       "[1163 rows x 10 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pepstats_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a3e083b",
   "metadata": {},
   "source": [
    "#### Normalise the Isoelectric point with standard scalar.. the others are already between 0 and 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "beb0cd29",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "cols_to_standardise = [\n",
    "    'Isoelectric Point'\n",
    "]\n",
    "\n",
    "scaler = StandardScaler()\n",
    "\n",
    "pepstats_normalised_df = pepstats_df.copy()\n",
    "\n",
    "pepstats_normalised_df[cols_to_standardise] = scaler.fit_transform(pepstats_normalised_df[cols_to_standardise])\n",
    "pepstats_normalised_df.to_csv(\"pepstats_results_standardised.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b502f6af",
   "metadata": {},
   "source": [
    "## Create the labels - known EC numbers of each UniProt ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "48da3c59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B7P3V8</td>\n",
       "      <td>2.7.7.49; 2.7.7.7; 3.1.26.4; 3.4.23.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D6VTK4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O13297</td>\n",
       "      <td>3.6.1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O13329</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O13516</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Entry                              EC number\n",
       "0  A0A0B7P3V8  2.7.7.49; 2.7.7.7; 3.1.26.4; 3.4.23.-\n",
       "1      D6VTK4                                    NaN\n",
       "2      O13297                               3.6.1.74\n",
       "3      O13329                                    NaN\n",
       "4      O13516                                    NaN"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ECnumbers_df = pd.read_csv('./databases/uniprot2EC_uniprotkb_download_2023_09_07.tsv', sep='\\t')\n",
    "ECnumbers_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b9c2baf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        Entry EC number\n",
      "0  A0A0B7P3V8       2.7\n",
      "1  A0A0B7P3V8       2.7\n",
      "2  A0A0B7P3V8       3.1\n",
      "3  A0A0B7P3V8       3.4\n",
      "5      O13297       3.6\n"
     ]
    }
   ],
   "source": [
    "ECnumbers_df = ECnumbers_df.assign(**{'EC number': ECnumbers_df['EC number'].str.split(';')}).explode('EC number').reset_index(drop=True)\n",
    "ECnumbers_df['EC number'] = ECnumbers_df['EC number'].str.strip()\n",
    "\n",
    "# Step 3: Discard everything beyond the second dot in the EC number\n",
    "ECnumbers_df['EC number'] = ECnumbers_df['EC number'].str.extract('(\\d+\\.\\d+)\\.')\n",
    "\n",
    "ECnumbers_df =ECnumbers_df.dropna(subset=['EC number'])\n",
    "\n",
    "# Display the first few rows of the modified dataframe\n",
    "print(ECnumbers_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dad4b109",
   "metadata": {},
   "source": [
    "## Combine all input variables and output class labels (EC numbers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "83f75547",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1093, 1281)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "\n",
    "# Load the sequence encodings\n",
    "with open(\"ESM2_encoded_representations.pkl\", \"rb\") as f:\n",
    "    encoded_representations = pickle.load(f)\n",
    "\n",
    "# Prepare data for DataFrame\n",
    "data = [(key,) + tuple(value.numpy()[0]) for key, value in encoded_representations.items()]\n",
    "\n",
    "# Prepare columns\n",
    "columns = ['Entry'] + [f'Encoding_{i+1}' for i in range(1280)]\n",
    "\n",
    "# Create dataframe\n",
    "ESMencoding_df = pd.DataFrame(data, columns=columns)\n",
    "ESMencoding_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c4d27fdf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>Isoelectric Point</th>\n",
       "      <th>Tiny</th>\n",
       "      <th>Small</th>\n",
       "      <th>Aliphatic</th>\n",
       "      <th>Aromatic</th>\n",
       "      <th>Non-polar</th>\n",
       "      <th>Polar</th>\n",
       "      <th>Charged</th>\n",
       "      <th>Basic</th>\n",
       "      <th>...</th>\n",
       "      <th>Encoding_1272</th>\n",
       "      <th>Encoding_1273</th>\n",
       "      <th>Encoding_1274</th>\n",
       "      <th>Encoding_1275</th>\n",
       "      <th>Encoding_1276</th>\n",
       "      <th>Encoding_1277</th>\n",
       "      <th>Encoding_1278</th>\n",
       "      <th>Encoding_1279</th>\n",
       "      <th>Encoding_1280</th>\n",
       "      <th>EC number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P33550</td>\n",
       "      <td>-0.940575</td>\n",
       "      <td>0.223529</td>\n",
       "      <td>0.421176</td>\n",
       "      <td>0.221176</td>\n",
       "      <td>0.185882</td>\n",
       "      <td>0.517647</td>\n",
       "      <td>0.482353</td>\n",
       "      <td>0.270588</td>\n",
       "      <td>0.134118</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006006</td>\n",
       "      <td>-0.078706</td>\n",
       "      <td>0.067504</td>\n",
       "      <td>-0.104536</td>\n",
       "      <td>0.024012</td>\n",
       "      <td>0.099171</td>\n",
       "      <td>-0.014483</td>\n",
       "      <td>0.063116</td>\n",
       "      <td>-0.103292</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q02207</td>\n",
       "      <td>1.390710</td>\n",
       "      <td>0.301111</td>\n",
       "      <td>0.521111</td>\n",
       "      <td>0.314444</td>\n",
       "      <td>0.113333</td>\n",
       "      <td>0.551111</td>\n",
       "      <td>0.448889</td>\n",
       "      <td>0.236667</td>\n",
       "      <td>0.136667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>-0.082104</td>\n",
       "      <td>0.138453</td>\n",
       "      <td>0.045217</td>\n",
       "      <td>0.080589</td>\n",
       "      <td>-0.004792</td>\n",
       "      <td>-0.045767</td>\n",
       "      <td>-0.085559</td>\n",
       "      <td>0.108033</td>\n",
       "      <td>1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q02207</td>\n",
       "      <td>1.390710</td>\n",
       "      <td>0.301111</td>\n",
       "      <td>0.521111</td>\n",
       "      <td>0.314444</td>\n",
       "      <td>0.113333</td>\n",
       "      <td>0.551111</td>\n",
       "      <td>0.448889</td>\n",
       "      <td>0.236667</td>\n",
       "      <td>0.136667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>-0.082104</td>\n",
       "      <td>0.138453</td>\n",
       "      <td>0.045217</td>\n",
       "      <td>0.080589</td>\n",
       "      <td>-0.004792</td>\n",
       "      <td>-0.045767</td>\n",
       "      <td>-0.085559</td>\n",
       "      <td>0.108033</td>\n",
       "      <td>4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P21375</td>\n",
       "      <td>1.417205</td>\n",
       "      <td>0.289421</td>\n",
       "      <td>0.520958</td>\n",
       "      <td>0.297405</td>\n",
       "      <td>0.083832</td>\n",
       "      <td>0.499002</td>\n",
       "      <td>0.500998</td>\n",
       "      <td>0.247505</td>\n",
       "      <td>0.143713</td>\n",
       "      <td>...</td>\n",
       "      <td>0.049613</td>\n",
       "      <td>-0.072365</td>\n",
       "      <td>0.065138</td>\n",
       "      <td>-0.225438</td>\n",
       "      <td>0.022500</td>\n",
       "      <td>-0.035332</td>\n",
       "      <td>-0.035813</td>\n",
       "      <td>0.007896</td>\n",
       "      <td>0.042926</td>\n",
       "      <td>1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P04046</td>\n",
       "      <td>-0.546599</td>\n",
       "      <td>0.284314</td>\n",
       "      <td>0.523529</td>\n",
       "      <td>0.290196</td>\n",
       "      <td>0.107843</td>\n",
       "      <td>0.547059</td>\n",
       "      <td>0.452941</td>\n",
       "      <td>0.268627</td>\n",
       "      <td>0.137255</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.129664</td>\n",
       "      <td>-0.186735</td>\n",
       "      <td>0.083157</td>\n",
       "      <td>-0.123854</td>\n",
       "      <td>-0.139891</td>\n",
       "      <td>0.068116</td>\n",
       "      <td>-0.059735</td>\n",
       "      <td>-0.066775</td>\n",
       "      <td>-0.084309</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1292 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniprotID  Isoelectric Point      Tiny     Small  Aliphatic  Aromatic  \\\n",
       "0    P33550          -0.940575  0.223529  0.421176   0.221176  0.185882   \n",
       "1    Q02207           1.390710  0.301111  0.521111   0.314444  0.113333   \n",
       "2    Q02207           1.390710  0.301111  0.521111   0.314444  0.113333   \n",
       "3    P21375           1.417205  0.289421  0.520958   0.297405  0.083832   \n",
       "4    P04046          -0.546599  0.284314  0.523529   0.290196  0.107843   \n",
       "\n",
       "   Non-polar     Polar   Charged     Basic  ...  Encoding_1272  Encoding_1273  \\\n",
       "0   0.517647  0.482353  0.270588  0.134118  ...       0.006006      -0.078706   \n",
       "1   0.551111  0.448889  0.236667  0.136667  ...       0.003671      -0.082104   \n",
       "2   0.551111  0.448889  0.236667  0.136667  ...       0.003671      -0.082104   \n",
       "3   0.499002  0.500998  0.247505  0.143713  ...       0.049613      -0.072365   \n",
       "4   0.547059  0.452941  0.268627  0.137255  ...      -0.129664      -0.186735   \n",
       "\n",
       "   Encoding_1274  Encoding_1275  Encoding_1276  Encoding_1277  Encoding_1278  \\\n",
       "0       0.067504      -0.104536       0.024012       0.099171      -0.014483   \n",
       "1       0.138453       0.045217       0.080589      -0.004792      -0.045767   \n",
       "2       0.138453       0.045217       0.080589      -0.004792      -0.045767   \n",
       "3       0.065138      -0.225438       0.022500      -0.035332      -0.035813   \n",
       "4       0.083157      -0.123854      -0.139891       0.068116      -0.059735   \n",
       "\n",
       "   Encoding_1279  Encoding_1280  EC number  \n",
       "0       0.063116      -0.103292        2.4  \n",
       "1      -0.085559       0.108033        1.1  \n",
       "2      -0.085559       0.108033        4.2  \n",
       "3       0.007896       0.042926        1.3  \n",
       "4      -0.066775      -0.084309        2.4  \n",
       "\n",
       "[5 rows x 1292 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Merge all three datasets based on the UniProtID\n",
    "combined_df = pepstats_normalised_df.merge(ESMencoding_df, left_index=True, right_on='Entry')\n",
    "combined_df = combined_df.merge(ECnumbers_df, on='Entry', how='inner')\n",
    "\n",
    "# Retain only those rows which have non-missing values in all columns\n",
    "combined_df = combined_df.dropna(how='any')\n",
    "\n",
    "# Rearrange such that the Entry column is the first, and it's called \"UniprotID\"\n",
    "combined_df.rename(columns={'Entry': 'UniprotID'}, inplace=True)\n",
    "cols = ['UniprotID'] + [col for col in combined_df.columns if col != 'UniprotID' and col != 'EC number'] + ['EC number']\n",
    "combined_df = combined_df[cols]\n",
    "\n",
    "combined_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "913235c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv(\"all_prepareddata_for_EzyPredict.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "529ce728",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(944, 1292)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (ML)",
   "language": "python",
   "name": "ml"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
